### ISSUE TITLE
Using pretrained embedding in Seq2seq with attention model

### ISSUE THREAD

**Issue by Sanghoon94** [tensorflow/models#1866]
Hello, 

I am trying to implement seq2seq model, by converting the sample code in 'models/tutorials/rnn/translate'.
It would be great if anyone can tell me how to use a pretrained embedding, such as google's word2vec.

It seems that the seq2seq model the sample code is using is "embedding_attention_seq2seq", which is from 'tf.contrib.legacy_seq2seq'.
Encoder is already attached with the embedding matrix using 'EmbeddingWrapper', and the decoder is "embedding_attention_decoder".

I don't want to train the embedding matrix while learning the seq2seq model, and is there any way to do so, while not building the seq2seq model on my own?

Thank you for your time.

**Comment by cy89** [comment]
@nealwu it looks like you own the models/tutorials/rnn models. Can you please advise @Sanghoon94 about the kind of pretraining they want to do?

**Comment by nealwu** [comment]
I think @ebrevdo would be the best person to help figure this out.

**Comment by ebrevdo** [comment]
You have to load the word2vec embeddings into the variable - and the order of rows in the loaded embeddings has to match (word-for-word/token-for-token) with the order of rows in the vocabulary you're using in your seq2seq model.

To do this; you can load the word2vec embeddings in python (if they fit in RAM) and run, via tf.session, a tf.assign() into the embeddings variable with the pretrained embedding values.  Again, make sure that each row k of the embedding variable corresponds to id k given by the vocabulary in the seq2seq problem.

Once you've done that; you need to modify wrap the output of embedding_lookup in a tf.stop_gradients to avoid backpropagating into it.

**Comment by PratsBhatt** [comment]
Hello @ebrevdo thank you for the answer. For the last point can it be done directly in the loop_function by removing the flag's if condition?
`
def loop_function(prev, _):
    if output_projection is not None:
      prev = nn_ops.xw_plus_b(prev, output_projection[0], output_projection[1])
    prev_symbol = math_ops.argmax(prev, 1)
    # Note that gradients will not propagate through the second parameter of
    # embedding_lookup.
    emb_prev = embedding_ops.embedding_lookup(embedding, prev_symbol)
    #if not update_embedding: # commenting the if clause so that it will always wrap it to stop_gradient?
    emb_prev = array_ops.stop_gradient(emb_prev)
    return emb_prev
`
thank you for your help and guidance.

**Comment by ebrevdo** [comment]
+rui

**Comment by oahziur** [comment]
@PratsBhatt Hi, I am not very familiar with your code, but I guess the loop function is only used by your decoder? 

I think you also need to use `tf.stop_gradient` for encoder's embedding inputs to avoid back-propagating into your embedding vars.

**Comment by PratsBhatt** [comment]
Hello @oahziur thanks a lot for your answer. The loop function is from the seq2seq.py from the legacy_seq2seq module. Yes it is used by the decoder. 
I have made a function that updates the embeddings by the assign function. 
1) Do I stop the embeddings at that point i.e. after the replacement or 
2) When I am building the graph I should make the `tf.stop_gradient` 
 
This is what I am thinking. 
seq2seq.py (legacy_seq2seq):
```
 # Encoder.
    encoder_cell = copy.deepcopy(cell)
    encoder_cell = core_rnn_cell.EmbeddingWrapper(
        encoder_cell,
        embedding_classes=num_encoder_symbols,
        embedding_size=embedding_size)
```
So stopping it it the EmbeddingWrapper class itself? like below 
```
 embedding = vs.get_variable(
          "embedding", [self._embedding_classes, self._embedding_size],
          initializer=initializer,
          dtype=data_type)
      embedded = embedding_ops.embedding_lookup(embedding,
                                                array_ops.reshape(inputs, [-1]))
      embedded = array_ops.stop_gradient(embedded)
```
Is my thinking correct. This way I would have to change the implementation of the tensorflow. Which I am not sure is a good idea. I do replace the embeddings. But not the embedding lookups. 
If this way is not good, then what if I `stop_gradient` on `embeddings` and not the `embedding_loopkup` object?
Sorry if the question is a little foolish, I am a beginner. Your help and guidance is valuable.
Thanks again.

**Comment by oahziur** [comment]
@PratsBhatt 

Hi, I think you can remove the EmbeddingWrapper, and do the embedding lookup & stop gradient in your seq2seq code. Then you can feed `embedded` as inputs directly when you run the encoder cell.

For example (with your example):
```
encoder_cell = copy.deepcopy(cell)
embedding = vs.get_variable(
          "embedding", [self._embedding_classes, self._embedding_size],
          initializer=initializer,
          dtype=data_type)
embedded = embedding_ops.embedding_lookup(embedding,array_ops.reshape(inputs, [-1]))                                          
embedded = array_ops.stop_gradient(embedded)
encoder_outputs, encoder_state = tf.nn.dynamic_rnn(encoder_cell, embedded, ...)
``` 

Hope that help.

**Comment by PratsBhatt** [comment]
Hi @oahziur , thank you for your answer. How do I check if it is working or not?
As per my understanding, the embeddings in the whole process should not change. i.e. after the training the embeddings should remain same as it were before when I had assign it?
Is my understanding correct? Thank you once again for all your guidance and help.

**Comment by oahziur** [comment]
Yes, the embeddings shouldn't change during the training process. I think you can eval the embedding variable after load a trained model to see if the it is same as your pretrained embeddings.

**Comment by PratsBhatt** [comment]
Thank you for your answer. One more question, what if I make the embeddings variable trainable=False when I create embeddings at the first place? Will that work out?

**Comment by oahziur** [comment]
@PratsBhatt Yes, I think set `trainable=False` should work as well. Make sure you use `tf.trainable_variables()` to get the `xs` if you calculate gradients yourself.

**Comment by sachinruk** [comment]
@PratsBhatt I'm assuming what you did was modify `embedding_attention_seq2seq` in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/legacy_seq2seq/python/ops/seq2seq.py . If so could you post your modified `embedding_attention_seq2seq`.

**Comment by tensorflowbutler** [comment]
Hi There,
 We are checking to see if you still need help on this, as this seems to be considerably old issue. Please update this issue with the latest information, code snippet to reproduce your issue and error you are seeing.
 If we don't hear from you in the next 7 days, this issue will be closed automatically. If you don't need help on this issue any more, please consider closing this.
